docs/docs/adrs/adr-015-partial-set-security.md

```diff
@@ -0,0 +1,308 @@
+Additionally, note that due to undelegations, etc. a validator that is in the top N% validators might fall down to the 100 - N% in which case this validator is not forced to validate anymore. Therefore, a validator that might move out of the top N% might want to send a `MsgOptIn` to keep validating even if this happens.
+
+Note that a validator can send a `MsgOptIn` message even if the consumer chain is not yet running. To do this we reuse the [`IsConsumerProposedOrRegistered`](https://github.com/cosmos/interchain-security/blob/v3.3.0/x/ccv/provider/keeper/key_assignment.go#L644). If the `chainID` does not exist, the `MsgOptIn` should fail, as well as if the provider address does not exist.
+
+Additionally, a validator that opts in can then use `MsgAssignConsumerKey` to change the consumer key on the consumer chain as is currently done in Replicated Security.
+
+#### State & Query
+We store the list of validators that have opted in and at the block height they opted it in under the key:
+```
+ChainIdWithLenKey(OptedInByteKey, chainID)
+```
+
+And for simplicity for each validator we store a boolean on whether this validator has opted in on this chain or not.
+```
+ChainIdAndConsAddrKey(OptedInByteKey, chainID, addr)
+```
+
+We augment the state of the provider to keep track for each chain all the validators that have opted in and if a specific validator has opted in.
+```
+// returns all the validators that have opted in on chain `chainID`
+func (k Keeper) GetOptedInValidators(ctx sdk.Context, chainID string) []Validators
+
+func (k Keeper) IsValidatorOptedIn(ctx sdk.Context, chainID string, val Validator) bool
+```
+
+We introduce a query to retrieve the validators that are opted in. The query would operate as follows.
+```
+gaiad query provider optedInValidators $chainId
+```
+
+#### When do validators opt in?
+Validators can not only opt in by sending a `MsgOptIn` message, but can also opt in automatically if they voted `Yes`
+during the `ConsumerAdditionProposal` that introduced a consumer chain. This simplifies validators operations because
+they do not have to send an additional message to opt in.
+
+Because the `Tally` method [deletes the votes](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/gov/keeper/tally.go#L71) after reading them, we cannot check the votes of the validators after the votes have been tallied. To circumvent this, we introduce
+a hook for [`AfterProposalVote`](https://github.com/cosmos/cosmos-sdk/blob/v0.47.7/x/gov/keeper/vote.go#L35) and keep track
+of all the votes cast by a validator. If a validator votes more than once, we only consider the latest vote.
+Finally, we only consider a validator has opted in if it casts a 100% `Yes` vote in case of a [weighted vote](https://github.com/cosmos/cosmos-sdk/blob/main/docs/architecture/adr-037-gov-split-vote.md).
+
+
+### How do validators opt out?
```

Comment:
To implement top-n, we should have an if condition here which checks if they are in the top-n and rejects their opt out message if they are.

Reply:
Added a sentence on this.